### ISSUE TITLE
defaultValue does not persist across form submissions

### ISSUE THREAD

**Issue by lifeiscontent** [edmundhung/conform#550]
### Describe the bug and the expected behavior

after form submissions defaultValues should be refreshed

### Conform version

v1.0.6

### Steps to Reproduce the Bug or Issue

Goto: https://stackblitz.com/edit/remix-run-remix-r2rkdc

# Expected

1. click toggle form
2. pick a value from the autocomplete list
3. hit submit
4. click toggle form
5. see that the value DOES persist

# Actual

Goto: https://stackblitz.com/edit/remix-run-remix-r2rkdc

1. click conform example
2. click toggle form
3. pick a value from the autocomplete list
4. hit submit
5. click toggle form
6. see that the value DOES NOT persist

### What browsers are you seeing the problem on?

Chrome

**Comment by edmundhung** [comment]
I can confirm that this does not happens on Firefox.

**Comment by lifeiscontent** [comment]
@edmundhung that was a false positive, because I didn't have autoComplete="off" on the input field, it was being prefilled with your autocomplete results, I've updated the reproduction with the necessary changes and now they work the same

**Comment by lifeiscontent** [comment]
@edmundhung do you know what part of the code this bug might exist? If so I could potentially create PR, but not sure where to look

**Comment by edmundhung** [comment]
This is partly related to #434. You need to delay when the form is reset so that Conform can pick up the latest default value on reset. (i.e. actionData is updated before loaderData), like this:

```tsx
const [form, fields] = useForm({
    lastResult: navigation.state === 'idle' ? lastResult : null,
    defaultValue: loaderData,
    onValidate({ formData }) {
      return parseWithZod(formData, { schema });
    },
    shouldValidate: 'onBlur',
});
```

However, adding this alone won't make your example work as well because the form is unmounted before reset happen and so Conform are not able to listen to the form reset event and act accordingly...

**Comment by lifeiscontent** [comment]
@edmundhung I'm guessing the conform handles some kind of submit event, why not rest defaultValue when lastResult changes since it already refreshes lastResult when it changes?

**Comment by edmundhung** [comment]
I am using DOM event as the source of truth for most the things happened in Conform. But I also made some exceptions recently because of all kinds of limitations I run into. I think I am fine to change this as well so that it reset the form directly instead of relying on the form reset event.

**Comment by edmundhung** [comment]
I can push up a fix tomorrow :)

**Comment by lifeiscontent** [comment]
@edmundhung fwiw the submit event could probably still be used here, honestly I wouldn't even mind if calling resetForm: true and THAT is how it updates the defaultValue, I think that would also be a reasonable expectation too.

**Comment by edmundhung** [comment]
It is. But I just don't see a viable solution right. Feel free to propose a solution if you have any idea!

**Comment by lifeiscontent** [comment]
@edmundhung could you maybe write up a small outline of how the internals of conform work starting from when I type something in, submit, and action comes back with lastResult?

that'd help me get up to speed and I can start digging in.

it would also help if you could touch on conform/dom and how that ties into the react lib

**Comment by edmundhung** [comment]
Sure! The code is messy and hard to follow 😅 So I hope this helps:

At a high level, Conform is basically a specialized Redux store with `FormData` as action and `metadata` as state. There are two reducers: one running in server action / `onValidate` that parse `FormData` into `Submission` through the [parse](https://github.com/edmundhung/conform/blob/ec101a2fb579e5438d443417a582c896bff050df/packages/conform-dom/submission.ts#L111-L225) helper), the other one running inside the `useForm` hook which populates the metadata based on the last submission result through the form context.

You can find the implementation of the form context in [this file](https://github.com/edmundhung/conform/blob/ec101a2fb579e5438d443417a582c896bff050df/packages/conform-dom/form.ts), especially the  `createFormContext` function. The form context does 3 things:

1. Maintain the raw `meta` (metadata) based on the submission result
2. Prepare the final `state` that includes proxy which derive complex metadata with caches 
3. When the raw `meta` is updated, it looks up what each subscriber needed (The `subject`) and compares the subscribed metadata on the new `state` and the old `state` one by one until it found something changed.

The react package are mostly about bridging the integration:
- Initialize a form context in the `useForm` hook
- Hooks to subscribe to the form context (useForm / useField / useFormMetadata)
- Pack the final `state` from the form context nicely and track all the metadata usage inside a proxy
- Helpers to deriving all the properties
- Makes custom input works more alike native input with the `useControl` hook

**Comment by lifeiscontent** [comment]
@edmundhung cool, I'm gonna put a test case together and see if I can work my way through it over the weekend.
